File: random-pick-with-weight/random-pick-with-weight.java
class Solution {
    int sum;
    List<Integer> rgs;
    Set<Integer> key;
    public Solution(int[] w) {
        this.sum = Arrays.stream(w).sum();
        this.rgs = new ArrayList<Integer>();
        int presum = 0;
        
        for(int i=0; i<w.length; i++){
            this.rgs.add(presum);
            presum += w[i];
        }
    }
    
    public int pickIndex() {
        int rnd = (int)(Math.random()*(this.sum));
        int l = 0, r = this.rgs.size();
        
        while(l+1<r){
            int mid = l + (r-l)/2;
            if(this.rgs.get(mid)==rnd){
                return mid;
            }else if(this.rgs.get(mid)>rnd){
                r = mid;
            }else if(this.rgs.get(mid)<rnd){
                l = mid;
            }
        }
        return l;
    }
}
/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */

File: 912-random-pick-with-weight/random-pick-with-weight.py
class Solution:

    def __init__(self, w: List[int]):
        sum_w = sum(w)
        self.n = len(w)

        # Get the PDF
        self.pdf = [float(v/sum_w) for v in w]

        # Get the CDF
        self.cdf = [0]*self.n
        self.cdf[0] = self.pdf[0]
        for idx in range(1, self.n):
            self.cdf[idx] = self.cdf[idx-1] + self.pdf[idx]


    def pickIndex(self) -> int:
        # Sample u from [0,1]
        random_number = random.random()

        # Return the smallest i such that cdf(x) >= u 
        return bisect.bisect(self.cdf, random_number)


    

# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()